src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/IO/TdsBufferAlloc.cs

```diff
@@ -0,0 +1,76 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+using System.Threading.Tasks;
+
+namespace Microsoft.Data.SqlClientX.IO
+{
+    /// <summary>
+    /// Provides buffer management for reader/writer implementations
+    /// </summary>
+    internal class TdsBufferAlloc : IDisposable, IAsyncDisposable
+    {
+        /// <summary>
+        /// Scratch buffer to serialize Short values (2 bytes).
+        /// </summary>
+        private byte[] _2Buffer;
+
+        /// <summary>
+        /// Scratch buffer to serialize Int values (4 bytes).
+        /// </summary>
+        private byte[] _4Buffer;
+
+        /// <summary>
+        /// Scratch buffer to serialize Long values (8 bytes).
+        /// </summary>
+        private byte[] _8Buffer;
+
+        /// <summary>
+        /// Scratch buffer to serialize Decimal values (16 bytes).
+        /// </summary>
+        private byte[] _16Buffer;
+
+        /// <summary>
+        /// Get buffer for reading/writing data
+        /// Accepted sizes are: 2, 4, 8, 16
+        /// </summary>
+        /// <param name="sizeInBytes">2, 4, 8 or 16 bytes of buffer size.</param>
+        /// <returns></returns>
+        internal byte[] GetBuffer(int sizeInBytes)
+        {
+            Debug.Assert(sizeInBytes == 2 || sizeInBytes == 4 || sizeInBytes == 8 || sizeInBytes == 16);
+            return sizeInBytes switch
```

Comment:
Depending how benchmarking goes, it might be worth encouraging this method to be inlined with `[MethodImpl(MethodImplOptions.AggressiveInlining)]`.

I'd also (very tentatively) think about returning the 16-byte array in all cases. It'll shave off 14 bytes from the heap, and we already constrain the buffer size via `AsSpan` and `AsMemory` downstream. Doing so also has knock-on effects which might mean we can discard the `TdsBufferAlloc` class and move the 16-byte scratch buffer directly onto `TdsWriter`.

Reply:
I'm not completely sure of the best practice here either. My rationale is that `sizeInBytes` will always be a constant value for every invocation of the same method (i.e. writing a ushort will always result in the value being 2.) I essentially want `GetBuffer` to be inlined into the calling method so that the JIT can see that this parameter's value is constant and eliminate the branch. If that happens the method calls on the hot path are essentially replaced by a single field access.

The two key questions for me are whether the JIT can see this already, and whether the overhead of making a method call and going through the case statement is significant enough to warrant the compiler hint. It needs a bit of benchmarking here (when the rest of SqlClientX is working end-to-end.)